### PR TITLE
Replay migrations on the created schema before copying data

### PR DESCRIPTION
**Pull request by VicDeo**
Ref https://github.com/owncloud/core/issues/27075#issuecomment-351673046
some  more fixes:

- answer  'n' to the 'continue' question was ignored
- added a new line after progressbar completion. Because this sucks:
```
oc_account_terms
    0 [>---------------------------]oc_accounts
 1/1 [============================] 100%oc_addressbookchanges
```
now it looks like this:
```
oc_account_terms
    0 [>---------------------------]
oc_accounts
 1/1 [============================] 100%
oc_addressbookchanges
```

### CODE REVIEW

**Comment by VicDeo** [comment]
@PVince81 it fixes the conversion but it needs to be polished and covered with tests

File: apps/files_external/appinfo/app.php [hunk 1]
```diff
@@ -35,7 +35,7 @@
 $appContainer = \OC_Mount_Config::$app->getContainer();
 
 $config = \OC::$server->getConfig();
-if ($config->getAppValue('core', 'enable_external_storage', 'no') === 'yes') {
+if (class_exists('OCA\Files\App') && $config->getAppValue('core', 'enable_external_storage', 'no') === 'yes') {
```
Comment (PVince81) [inline apps/files_external/appinfo/app.php]: unrelated ? can you submit this as a separate tech debt PR ?

Comment (VicDeo) [reply apps/files_external/appinfo/app.php]: @PVince81 
Related. I need to load app to read its migrations. If `files` app is not loaded (it doesn't use migrations) - result of loading `files_external` is  :boom:

File: core/Command/Db/ConvertType.php [hunk 2]
```diff
@@ -195,67 +246,109 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
+	 */
+	protected function createSchema(Connection $fromDB, Connection $toDB, InputInterface $input, OutputInterface $output) {
 		$output->writeln('<info>Creating schema in new database</info>');
-		$schemaManager = new \OC\DB\MDB2SchemaManager($toDB);
+		$schemaManager = new MDB2SchemaManager($toDB);
 		$schemaManager->createDbFromStructure(\OC::$SERVERROOT.'/db_structure.xml');
-		$apps = $input->getOption('all-apps') ? \OC_App::getAllApps() : \OC_App::getEnabledApps();
+		$this->replayMigrations($fromDB, $toDB, 'core');
+
+		$apps = $input->getOption('all-apps') ? $this->appManager->getAllApps() : $this->appManager->getInstalledApps();
```
Comment (VicDeo) [inline core/Command/Db/ConvertType.php]: `getAllApps` is not what we need here

File: core/Command/Db/ConvertType.php [hunk 3]
```diff
@@ -195,67 +246,126 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
+	 * @throws \OC\NeedsUpdateException
+	 */
+	protected function replayMigrations(Connection $fromDB, Connection $toDB, $app) {
+		if ($app !== 'core') {
+			\OC_App::loadApp($app);
+		}
+		$sourceMigrationService = new MigrationService($app, $fromDB);
+		$currentMigration = $sourceMigrationService->getMigration('current');
+		if ($currentMigration !== '0') {
+			$targetMigrationService = new MigrationService($app, $toDB);
```
Comment (PVince81) [inline core/Command/Db/ConvertType.php]: are you not filtering migrations by "SQL" only ?

Comment (PVince81) [reply core/Command/Db/ConvertType.php]: Consider that some migrations are operating on data, for example making some conversion.

When converting the DB we don't need to re-process the data as it was already processed.

This only works if all migrations work correctly with already converted data and have no side effect.
We should likely make it a requirement that every migration must be "replayable" in a way that it doesn't break if the prerequisite was already been resolved before (ex: already migrated data)

@DeepDiver1975

Comment (VicDeo) [reply core/Command/Db/ConvertType.php]: @PVince81 target `migrations` table will not have these migrations if I skip them

Comment (PVince81) [reply core/Command/Db/ConvertType.php]: Good point. In any case it seems sensible to make sure that non-schema migrations work correctly when replayed

**Comment by PVince81** [comment]
what's left to do ?

**Comment by VicDeo** [comment]
@PVince81 in general it's ready but a functional test is nice-to-have here.

**Comment by PVince81** [comment]
I don't think QA has time to test this for 10.0.8 so we need to defer.

Another option would be to ship this as experimental and ask for help to test this. Experimental might be better than completely broken as it currently is on 10.0. Would require a warning message on the CLI.

Thoughts ? @pmaier1 @patrickjahns

**Comment by ichtyp** [comment]
I would go for the experimental option and test it at home.

**Comment by VicDeo** [comment]
@PVince81 I'm fine with any decision, just let me know if I need to add a warning

**Comment by macgeneral** [comment]
@PVince81

I'm not sure if I'm doing this correctly but it doesn't work for me:
* MySQL/MariaDB 10.1.32 -> PostgreSQL 10.3
* owncloud 10.0.7.2 with patch 30643 applied
* CentOS 7.4.1708

(Or did you only fix SQLite -> MySQL so far?)

> $ httpdocs.patched: patch -p1 < ../30643.patch
> 
> patching file core/Command/Db/ConvertType.php
> patching file apps/files_external/appinfo/app.php
> patching file apps/systemtags/appinfo/app.php
> Hunk #1 FAILED at 77.
> 1 out of 1 hunk FAILED -- saving rejects to file apps/systemtags/appinfo/app.php.rej
> patching file core/Command/Db/ConvertType.php
> patching file core/register_command.php
> Hunk #1 succeeded at 82 (offset -1 lines).
> patching file tests/Core/Command/Db/ConvertTypeTest.php
> patching file core/Command/Db/ConvertType.php
> patching file tests/Core/Command/Db/ConvertTypeTest.php`


> cat  apps/systemtags/appinfo/app.php.rej
> --- apps/systemtags/appinfo/app.php
> +++ apps/systemtags/appinfo/app.php
> @@ -77,13 +77,16 @@
>  $eventDispatcher->addListener(MapperEvent::EVENT_ASSIGN, $mapperListener);
>  $eventDispatcher->addListener(MapperEvent::EVENT_UNASSIGN, $mapperListener);
> 
> -\OCA\Files\App::getNavigationManager()->add(function () {
> -	$l = \OC::$server->getL10N('systemtags');
> -	return [
> -		'id' => 'systemtagsfilter',
> -		'appname' => 'systemtags',
> -		'script' => 'list.php',
> -		'order' => 25,
> -		'name' => $l->t('Tags')
> -	];
> -});
> +if (class_exists('OCA\Files\App')) {
> +	\OCA\Files\App::getNavigationManager()->add(function () {
> +		$l = \OC::$server->getL10N('systemtags');
> +		return [
> +			'id' => 'systemtagsfilter',
> +			'appname' => 'systemtags',
> +			'script' => 'list.php',
> +			'order' => 25,
> +			'name' => $l->t('Tags')
> +		];
> +	});
> +}
> +


and then:
> sudo -u webuser php occ db:convert-type --all-apps pgsql owncloud /var/run/postgresql owncloud

creates the tables (empty) and fails with:

> Creating schema in new database
> 
> 
>   [Doctrine\DBAL\Exception\InvalidFieldNameException]
>   An exception occurred while executing 'SELECT min_value, increment_by FROM "oc_storages_numeric_id_seq"':
>   SQLSTATE[42703]: Undefined column: 7 ERROR:  column "min_value" does not exist
>   LINE 1: SELECT min_value, increment_by FROM "oc_storages_numeric_id_...
>                  ^
> 
> 
> 
>   [Doctrine\DBAL\Driver\PDOException]
>   SQLSTATE[42703]: Undefined column: 7 ERROR:  column "min_value" does not exist
>   LINE 1: SELECT min_value, increment_by FROM "oc_storages_numeric_id_...
>                  ^
> 
> 
> 
>   [PDOException]
>   SQLSTATE[42703]: Undefined column: 7 ERROR:  column "min_value" does not exist
>   LINE 1: SELECT min_value, increment_by FROM "oc_storages_numeric_id_...
>                  ^

**Comment by VicDeo** [comment]
@macgeneral regarding postgre 10:
 https://github.com/owncloud/core/issues/29342

**Comment by macgeneral** [comment]
@VicDeo:
Ok with the patch from @wzator: https://github.com/doctrine/dbal/issues/2868#issuecomment-332119007 and after increasing the php cli memory it seems to work so far!

I will create a new vhost tomorrow running with Postgres to test if everything is there as expected:
`$ httpdocs.patched: sudo -u webuser php occ db:convert-type --all-apps pgsql owncloud /var/run/postgresql owncloud`

> Enter a password to access a target database: [removed]
> Creating schema in new database
> The following tables will not be converted:
> oc_bookmarks
> oc_bookmarks_tags
> oc_contacts_cards_properties
> oc_documents_invite
> oc_documents_member
> oc_documents_op
> oc_documents_revisions
> oc_documents_session
> oc_mail_accounts
> oc_mail_collected_addresses
> oc_mozilla_sync_collections
> oc_mozilla_sync_users
> oc_mozilla_sync_wbo
> oc_oc_filecache_j9dbhe2stb4wv
> oc_ownnote
> oc_ownnote_parts
> oc_richdocuments_invite
> oc_richdocuments_member
> oc_richdocuments_op
> oc_richdocuments_revisions
> oc_richdocuments_session
> oc_storagecharts2
> oc_storagecharts2_uconf
> oc_trusted_servers
> Continue with the conversion (y/n)? [n] y
> oc_appconfig
>  241/241 [============================] 100%
> oc_storages
>  16/16 [============================] 100%
> oc_mounts
>  11/11 [============================] 100%
> oc_mimetypes
>  125/125 [============================] 100%
> oc_group_user
>  12/12 [============================] 100%
> oc_group_admin
>  1/1 [============================] 100%
> oc_groups
>  3/3 [============================] 100%
> oc_preferences
>  102/102 [============================] 100%
> oc_users
>  3/3 [============================] 100%
> oc_jobs
>  25/25 [============================] 100%
> oc_authtoken
>  33/33 [============================] 100%
> oc_vcategory
>  45/45 [============================] 100%
> oc_vcategory_to_object
>     0 [>---------------------------]
> oc_systemtag
>     0 [>---------------------------]
> oc_systemtag_object_mapping
>     0 [>---------------------------]
> oc_systemtag_group
>     0 [>---------------------------]
> oc_privatedata
>     0 [>---------------------------]
> oc_file_locks
>     0 [>---------------------------]
> oc_comments
>     0 [>---------------------------]
> oc_comments_read_markers
>     0 [>---------------------------]
> oc_credentials
>     0 [>---------------------------]
> Skipping "oc_migrations" as it already has migrations data
> oc_addressbookchanges
>  878/878 [============================] 100%
> oc_addressbooks
>  12/12 [============================] 100%
> oc_calendarchanges
> chunked query, 5 chunks
>  4672/4672 [============================] 100%
> oc_dav_shares
>  8/8 [============================] 100%
> oc_calendarsubscriptions
>     0 [>---------------------------]
> oc_cards
>  611/611 [============================] 100%
> oc_cards_properties
> chunked query, 4 chunks
>  3427/3427 [============================] 100%
> oc_schedulingobjects
>     0 [>---------------------------]
> oc_calendarobjects
> chunked query, 3 chunks
>  2380/2380 [============================] 100%
> oc_calendars
>  27/27 [============================] 100%
> oc_external_mounts
>     0 [>---------------------------]
> oc_external_applicable
>     0 [>---------------------------]
> oc_external_config
>     0 [>---------------------------]
> oc_external_options
>     0 [>---------------------------]
> oc_accounts
>  5/5 [============================] 100%
> oc_account_terms
>     0 [>---------------------------]
> oc_share
>  4/4 [============================] 100%
> oc_activity
>  605/605 [============================] 100%
> oc_filecache
> chunked query, 140 chunks
>  139442/139442 [============================] 100%
> oc_activity_mq
>     0 [>---------------------------]
> oc_audioplayer_artists
>  1/1 [============================] 100%
> oc_audioplayer_albums
>  1/1 [============================] 100%
> oc_audioplayer_genre
>  1/1 [============================] 100%
> oc_audioplayer_playlists
>     0 [>---------------------------]
> oc_audioplayer_playlist_tracks
>     0 [>---------------------------]
> oc_audioplayer_tracks
>  1/1 [============================] 100%
> oc_audioplayer_streams
>     0 [>---------------------------]
> oc_audioplayer_stats
>  1/1 [============================] 100%
> oc_properties
> chunked query, 3 chunks
>  2885/2885 [============================] 100%
> oc_dav_properties
>  4/4 [============================] 100%
> oc_federated_reshares
>     0 [>---------------------------]
> oc_files_antivirus
> chunked query, 2 chunks
>  1174/1174 [============================] 100%
> oc_files_avir_status
>  19/19 [============================] 100%
> oc_share_external
>     0 [>---------------------------]
> oc_files_trash
>     0 [>---------------------------]
> oc_notifications
>     0 [>---------------------------]
> oc_polls_events
>     0 [>---------------------------]
> oc_polls_dts
>     0 [>---------------------------]
> oc_polls_txts
>     0 [>---------------------------]
> oc_polls_particip
>     0 [>---------------------------]
> oc_polls_particip_text
>     0 [>---------------------------]
> oc_polls_comments
>     0 [>---------------------------]
> oc_polls_notif
>     0 [>---------------------------]
> oc_richdocuments_wopi
>  129/129 [============================] 100%
> oc_twofactor_totp_secrets
>  1/1 [============================] 100%
> oc_users_external
>  8/8 [============================] 100%
>

**Comment by ichtyp** [comment]
> However if people want to help testing they could download this PR as patch 
> https://github.com/owncloud/core/pull/30643.patch ( @ichtyp ) 

@PVince81 , as I installed OC via the package management from FreeNAS, can i just take the four files and copy them over, or will I break something ? :-)

**Comment by PVince81** [comment]
PHP CS Fixer needs your attention

**Comment by VicDeo** [comment]
addressed PHP CS failure and squashed a bit

**Comment by VicDeo** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by ownclouders** [comment]
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

**Comment by VicDeo** [comment]
conflicts are resolved

**Comment by rienex** [comment]
Hi,

is there special think to do before?
I have latest Version 10.0.9 stable in an qemu virtualisation installed (ubuntu server 16.04 latest)

I try it in that way:

```
cd /var/www/owncloud/
sudo -u www-data ./occ maintenance:mode --on
sudo service apache2 stop

sudo patch -p1 < 30643.patch
patching file core/Command/Db/ConvertType.php
Hunk #3 FAILED at 150.
Hunk #5 FAILED at 191.
Hunk #7 FAILED at 243.
Hunk #9 succeeded at 313 (offset 1 line).
Hunk #10 FAILED at 348.
Hunk #11 succeeded at 379 (offset 1 line).
4 out of 11 hunks FAILED -- saving rejects to file core/Command/Db/ConvertType.php.rej
patching file core/register_command.php
Hunk #1 succeeded at 82 (offset -2 lines).
patching file tests/Core/Command/Db/ConvertTypeTest.php
patching file apps/files_external/appinfo/app.php
Hunk #1 succeeded at 35 (offset 2 lines).
patching file apps/systemtags/appinfo/app.php

```

I can provide the ConvertType.php.rej if you are interested ?

Thank you

regards Tino

**Comment by DJGummikuh** [comment]
I double @rienex observation, I also can't patch

**Comment by grauth** [comment]
I patched today and at least the oc_jobs table has now no issues anymore. But I stuck at oc_mimetypes.

In AbstractMySQLDriver.php line 66:
  An exception occurred while executing 'INSERT INTO `oc_mimetypes` (`id`, `mimetype`) VALUES(?, ?)' with params ["74", "video\/MP2T"]:
  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'video/MP2T' for key 'mimetype_id_index'
In PDOStatement.php line 107:
  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'video/MP2T' for key 'mimetype_id_index'

In PDOStatement.php line 105:
  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'video/MP2T' for key 'mimetype_id_index'

occ status
  - installed: true
  - first_install_version: unknown
  - version: 10.0.10.4
  - versionstring: 10.0.10
  - edition: Community

**Comment by PVince81** [comment]
@VicDeo would this be safe to merge and doesn't affect other code paths ?

am thinking to release this as beta, which is a step forward from "broken" and could help with testing (no wild patching)

**Comment by VicDeo** [comment]
@grauth could you please share the entries from your `oc_mimetypes` table?

**Comment by VicDeo** [comment]
@PVince81 
> would this be safe to merge and doesn't affect other code paths?

No, it doesn't affect other code paths.

**Comment by PVince81** [comment]
I'm merging this as it would bring the command in a more usable state, potentially with remaining bugs to uncover. We should advertise this as beta / experimental for now until more testing is done.

@pmaier1 FYI

**Comment by PVince81** [comment]
@VicDeo can you add a note on the CLI help / output that this is currently experimental ?

**Comment by TobeStokes** [comment]
I'm trying to follow this and figure out what the status is for the database conversion tool...   I have OC 10.0.10.4 installed but unfortunately using SQLite and want to convert to MariaDB.  I found out the hard way about the table limit in SQLite last December.  Managed to recover and stabilize everything, but really want to get away from SQLite.

Is the DB conversion tool usable in 10.0.10 yet?

**Comment by VicDeo** [comment]
@PVince81  > can you add a note on the CLI help / output that this is currently experimental ?

done

@TobeStokes 
>Is the DB conversion tool usable in 10.0.10 yet?

No, it isn't

**Comment by VicDeo** [comment]
@jvillafanez thanks for your efforts but I have no plans to invest more time into this 1 year old PR until it will be planned for the sprint

**Comment by jvillafanez** [comment]
I think you're using the class as a data holder (for the DB parameters)... not sure if we have time to improve that.

File: core/Command/Db/ConvertType.php [hunk 4]
```diff
@@ -152,34 +193,46 @@ protected function readPassword(InputInterface $input, OutputInterface $output)
...
+	 * @return int|null|void
+	 * @throws \Exception
+	 */
 	protected function execute(InputInterface $input, OutputInterface $output) {
-		$this->validateInput($input, $output);
-		$this->readPassword($input, $output);
+		$output->writeln('<info>This feature is currently experimental.</info>');
+		$this->targetType = $this->connectionFactory->normalizeType($input->getArgument('type'));
+		$this->targetHostname = $input->getArgument('hostname');
+		$this->targetPort = $input->getOption('port') ? $input->getOption('port') : null;
```
[the remark this answers is not in the document]
Comment (VicDeo) [reply core/Command/Db/ConvertType.php]: there is no problem except I did't want to pass `$input` into all methods that need to know the  port number.
What could return if there is no port? `NULL` (and this is valid for sqlite)

**Comment by PVince81** [comment]
we should at least finish and ship as beta. for that we need a review.
if it turns out that adjusting it takes more time, we can schedule this again.

File: core/Command/Db/ConvertType.php [hunk 5]
```diff
@@ -272,12 +385,25 @@ protected function copyTable(Connection $fromDB, Connection $toDB, $table, Input
 		$query = $fromDB->getQueryBuilder();
 		$query->automaticTablePrefix(false);
 		$query->select('*')
-			->from($table)
+			->from($tableName)
 			->setMaxResults($chunkSize);
 
+		try {
+			$orderColumns = $table->getPrimaryKeyColumns();
```
[the remark this answers is not in the document]
Comment (VicDeo) [reply core/Command/Db/ConvertType.php]: because primary key is faster (but could be missing)

[the remark this answers is not in the document]
Comment (VicDeo) [reply core/Command/Db/ConvertType.php]: We are copying data from one database to another and should enforce an order for SELECT. Primary key is the best option in this case. But there can be tables without it.

**Comment by pmaier1** [comment]
> we should at least finish and ship as beta.

Yes, if you feel confident now, let's include it in the next minor and mention that it's beta / waiting for feedback.

File: core/Command/Db/ConvertType.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (VicDeo) [reply core/Command/Db/ConvertType.php]: changed to `$this->targetPort = $input->getOption('port')`

File: core/Command/Db/ConvertType.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (VicDeo) [reply core/Command/Db/ConvertType.php]: comments have been added

File: core/Command/Db/ConvertType.php [hunk 6]
```diff
@@ -295,52 +421,110 @@ protected function copyTable(Connection $fromDB, Connection $toDB, $table, Input
...
+			$fromSchema = $fromDB->createSchema();
 			// copy table rows
-			foreach ($tables as $table) {
-				$output->writeln($table);
+			foreach ($tables as $tableName) {
+				$table = $fromSchema->getTable($tableName);
+				if ($tableName === $toDB->getPrefix() . 'migrations') {
+					$output->writeln(
+						\sprintf(
+							'<info>Skipping "%s" as it already has migrations data</info>',
```
[the remark this answers is not in the document]
Comment (VicDeo) [reply core/Command/Db/ConvertType.php]: changed to `'<info>Skipping copying data for the table "%s", it will be populated later.</info>',`

**Comment by VicDeo** [comment]
@jvillafanez please rereview

**Comment by jvillafanez** [comment]
How much time do we have or do we want to invest here? There are several changes we should consider to improve the code quality, but those will take some time.
I haven't checked in depth the possibilities of unittests in the command, but I'm pretty sure it isn't fully covered with just 3 tests, plus the tests aren't isolated: the MDB2SchemaManager and MigrationService classes aren't mocked.

**Comment by michaelstingl** [comment]
@micbar @pmaier1 Change milestone to 10.3?

**Comment by pmaier1** [comment]
> How much time do we have or do we want to invest here? 

It should work and not cause issues, of course. Anything else is out-of-scope here right now. If that's satisfied here, let's ship this with the next minor release and let people try and get feedback.

@micbar can you take this into your scope for the next release, please? THX.

**Review by jvillafanez — APPROVED** [review]
I assume the code works properly.
There are several code changes we should do, but we can't in the allocated time

**Comment by VicDeo** [comment]
brifly retested and rebased once again

**Comment by VicDeo** [comment]
stable10: https://github.com/owncloud/core/pull/35390

**Pull request merged**